homework/shared_ptr/weak_ptr.hpp

```diff
@@ -75,29 +71,55 @@ weak_ptr<T>::~weak_ptr() {
 }
 
 template <typename T>
-weak_ptr<T>& weak_ptr<T>::operator=(const weak_ptr<T>& ptr) noexcept {
+weak_ptr<T>& weak_ptr<T>::operator=(const weak_ptr& ptr) noexcept {
+    ptr_ = ptr.ptr_;
+    counter_ = ptr.counter_;
     if (counter_) {
-        counter_->decreaseWeakRefs();
+        counter_->increaseWeakRefs();
     }
-    //checkAndDeletePointers();
-    counter_ = ptr.counter_;
-    ptr_ = ptr.ptr_;
-    counter_->increaseWeakRefs();
+    return *this;
+}
 
+template <typename T>
+weak_ptr<T>& weak_ptr<T>::operator=(const shared_ptr<T>& ptr) noexcept {
+    ptr_ = ptr.ptr_;
+    counter_ = ptr.counter_;
+    if (counter_) {
+        counter_->increaseWeakRefs();
+    }
     return *this;
 }
 
+
 template <typename T>
-weak_ptr<T>& weak_ptr<T>::operator=(weak_ptr<T>&& previousOwner) noexcept {
+weak_ptr<T>& weak_ptr<T>::operator=(weak_ptr&& previousOwner) noexcept {
     if (this != &previousOwner) {
-        counter_->decreaseWeakRefs();
-        //checkAndDeletePointers();
-        ptr_ = previousOwner.ptr_;
-        counter_ = previousOwner.counter_;
+        weak_ptr(previousOwner).swap(*this);
         previousOwner.ptr_ = nullptr;
         previousOwner.counter_ = nullptr;
     }
     return *this;
 }
 
+template <typename T>
+void weak_ptr<T>::reset() noexcept {
+    weak_ptr{}.swap(*this);
```

Comment:
This will lead to memory leak. You don't decrease weak pointer count and do not check if memory of `counter_` is to be released.

Reply:
done